File: ios/PTimerKit/Sources/PTimerCore/Exposure/FilterInventory.swift
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Stable identity of a user-owned Filter Set. Names and colors are
/// presentation metadata; only this id participates in stack
/// references and captured summaries.
public struct FilterSetID: Hashable, Codable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }

    public static func generate() -> FilterSetID {
        FilterSetID(rawValue: UUID().uuidString)
    }
}

/// Stable identity of one physical filter. Two items with equal
/// names and values remain two distinct physical filters.
public struct FilterItemID: Hashable, Codable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }

    public static func generate() -> FilterItemID {
        FilterItemID(rawValue: UUID().uuidString)
    }
}

/// Fixed, platform-neutral color palette for Filter Sets. Persisted
/// by `rawValue` so both platforms map the same token to their own
/// color system. Duplicate colors across Filter Sets are allowed.
public enum FilterSetColor: String, Codable, CaseIterable, Sendable {
    case red
    case orange
    case yellow
    case green
    case mint
    case teal
    case cyan
    case blue
    case indigo
    case purple
    case pink
    case brown

    /// Random suggestion for a new Filter Set that differs from the
    /// suggestion offered on the immediately preceding creation
    /// opening. With one color excluded there are always eleven
    /// candidates left, so the call never fails.
    public static func suggestion<G: RandomNumberGenerator>(
        excluding previous: FilterSetColor?,
        using generator: inout G
    ) -> FilterSetColor {
        let candidates = allCases.filter { $0 != previous }
        return candidates.randomElement(using: &generator) ?? .blue
    }

    public static func suggestion(excluding previous: FilterSetColor?) -> FilterSetColor {
        var generator = SystemRandomNumberGenerator()
        return suggestion(excluding: previous, using: &generator)
    }
}

/// Input unit of a registered filter value. Conversion to canonical
/// stops: Stops unchanged, `OD / 0.3`, `log2(ND factor)`.
public enum FilterValueUnit: String, Codable, CaseIterable, Sendable {
    case stops
    case opticalDensity
    case filterFactor
}

/// A registered Fixed or GND filter value as the user entered it —
/// the original value and unit are preserved for display, while
/// calculation consumes `canonicalStops`.
public struct FilterRegisteredValue: Hashable, Codable, Sendable {
    public let value: Double
    public let unit: FilterValueUnit

    public init(value: Double, unit: FilterValueUnit) {
        self.value = value
        self.unit = unit
    }

    /// Canonical stops, or `nil` when the value cannot be registered:
    /// non-finite, zero or negative stops, or above the 30-stop cap.
    /// The result is never snapped to the Standard ladder.
    public var canonicalStops: Double? {
        Self.canonicalStops(value: value, unit: unit)
    }

    public var isValid: Bool {
        canonicalStops != nil
    }

    public static func canonicalStops(value: Double, unit: FilterValueUnit) -> Double? {
        guard value.isFinite else {
            return nil
        }
        let stops: Double
        switch unit {
        case .stops:
            stops = value
        case .opticalDensity:
            stops = value / 0.3
        case .filterFactor:
            stops = value > 0 ? log2(value) : -.infinity
        }
        guard stops.isFinite,
              stops > 0,
              stops <= Double(ExposureScale.maximumWholeNDStops) + ExposureCalculator.stabilityEpsilon else {
            return nil
        }
        return stops
    }
}

/// Parses user-entered decimal text for filter registration. Accepts
/// either `.` or `,` as the decimal separator so locale keyboards,
/// pasted text, and hardware keyboards normalize to one rule.
public enum FilterDecimalInput {
    public static func normalizedText(_ text: String) -> String {
        text.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: ",", with: ".")
    }

    /// General decimal parse (Fixed / GND values). `nil` for empty or
    /// non-numeric text.
    public static func parseDecimal(_ text: String) -> Double? {
        let normalized = normalizedText(text)
        guard !normalized.isEmpty,
              normalized.range(of: "^[0-9]+(\\.[0-9]+)?$|^\\.[0-9]+$", options: .regularExpression) != nil,
              let value = Double(normalized) else {
            return nil
        }
        return value
    }

    /// CPL exposure-loss field parse: one integer digit, at most one
    /// fractional digit, in the closed range 0.1–9.9. `.empty` for
    /// blank text, `.invalid` for anything else.
    public enum CPLFieldParse: Equatable, Sendable {
        case empty
        case value(Double)
        case invalid
    }

    public static func parseCPLField(_ text: String) -> CPLFieldParse {
        let normalized = normalizedText(text)
        guard !normalized.isEmpty else {
            return .empty
        }
        guard normalized.range(of: "^[0-9](\\.[0-9])?$", options: .regularExpression) != nil,
              let value = Double(normalized),
              CPLExposureLossChoices.isValidChoice(value) else {
            return .invalid
        }
        return .value(value)
    }
}

/// The three user-editable CPL exposure-loss choices. A field may be
/// empty (omits a choice); at least one field must hold a valid value.
public struct CPLExposureLossChoices: Hashable, Codable, Sendable {
    public static let fieldCount = 3
    public static let range: ClosedRange<Double> = 0.1...9.9
    /// Shipping defaults for a new CPL item: 1, 1.5, and 2 stops.
    public static let defaults = CPLExposureLossChoices(fields: [1, 1.5, 2])

    /// Exactly three slots; `nil` is an empty field.
    public let fields: [Double?]

    public init(fields: [Double?]) {
        var padded = Array(fields.prefix(Self.fieldCount))
        while padded.count < Self.fieldCount {
            padded.append(nil)
        }
        self.fields = padded
    }

    /// A choice is valid when it is finite, inside 0.1–9.9, and has
    /// at most one fractional digit (so `1.25` is rejected).
    public static func isValidChoice(_ value: Double) -> Bool {
        guard value.isFinite, range.contains(value) else {
            return false
        }
        let tenths = value * 10
        return abs(tenths - tenths.rounded()) <= ExposureCalculator.stabilityEpsilon
    }

    public var isValid: Bool {
        let present = fields.compactMap { $0 }
        return !present.isEmpty && present.allSatisfy(Self.isValidChoice)
    }

    /// Distinct valid choices in ascending order — the rows the
    /// shooting wheel offers for this item. Duplicates appear once.
    public var shootingChoices: [Double] {
        var seen: [Double] = []
        for value in fields.compactMap({ $0 }) where Self.isValidChoice(value) {
            if !seen.contains(where: { abs($0 - value) <= ExposureCalculator.stabilityEpsilon }) {
                seen.append(value)
            }
        }
        return seen.sorted()
    }
}

/// Per-shot GND calculation mode.
public enum GNDCalculationMode: String, Codable, CaseIterable, Sendable {
    /// The GND is mounted and recorded; it contributes 0 stops.
    case recordOnly
    /// The GND contributes its complete registered value.
    case applyFullValue
}

/// Behavior kind of a physical item — chosen explicitly by the user,
/// never inferred from the name.
public enum FilterItemKind: String, Codable, CaseIterable, Sendable {
    case fixed
    case cpl
    case gnd
}

public enum FilterItemBehavior: Hashable, Sendable {
    case fixed(FilterRegisteredValue)
    case cpl(CPLExposureLossChoices)
    case gnd(FilterRegisteredValue)

    public var kind: FilterItemKind {
        switch self {
        case .fixed: return .fixed
        case .cpl: return .cpl
        case .gnd: return .gnd
        }
    }

    /// Registered full-density value for Fixed and GND items; CPL
    /// items carry choices instead.
    public var registeredValue: FilterRegisteredValue? {
        switch self {
        case .fixed(let value), .gnd(let value):
            return value
        case .cpl:
            return nil
        }
    }

    public var isValid: Bool {
        switch self {
        case .fixed(let value), .gnd(let value):
            return value.isValid
        case .cpl(let choices):
            return choices.isValid
        }
    }
}

/// One physical filter with a stable id. Equal names, kinds, units,
/// and values do not make two items the same item.
public struct FilterItem: Identifiable, Hashable, Sendable {
    public let id: FilterItemID
    public var name: String
    public var behavior: FilterItemBehavior

    public init(id: FilterItemID = .generate(), name: String, behavior: FilterItemBehavior) {
        self.id = id
        self.name = name
        self.behavior = behavior
    }

    public var isWellFormed: Bool {
        !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && behavior.isValid
    }
}

/// A user-owned, ordered group of physical filters with a stable id,
/// a non-empty name, and a required color.
public struct FilterSet: Identifiable, Hashable, Sendable {
    public let id: FilterSetID
    public var name: String
    public var color: FilterSetColor
    public var items: [FilterItem]

    public init(id: FilterSetID = .generate(), name: String, color: FilterSetColor, items: [FilterItem] = []) {
        self.id = id
        self.name = name
        self.color = color
        self.items = items
    }

    public var isWellFormed: Bool {
        !name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    public func item(withID id: FilterItemID) -> FilterItem? {
        items.first { $0.id == id }
    }
}

/// The complete user inventory: Filter Sets in user-defined display
/// order. Standard is a fixed built-in source that always precedes
/// every Filter Set.
public struct FilterInventory: Hashable, Sendable {
    public var filterSets: [FilterSet]

    public init(filterSets: [FilterSet] = []) {
        self.filterSets = filterSets
    }

    public static let empty = FilterInventory()

    public func filterSet(withID id: FilterSetID) -> FilterSet? {
        filterSets.first { $0.id == id }
    }

    public func item(withID id: FilterItemID) -> (filterSet: FilterSet, item: FilterItem)? {
        for filterSet in filterSets {
            if let item = filterSet.item(withID: id) {
                return (filterSet, item)
            }
        }
        return nil
    }

    /// Filter Sources in selection order: Standard, then Filter Sets
    /// in their user-defined order.
    public var sources: [FilterSource] {
        [.standard] + filterSets.map { .filterSet($0.id) }
    }

    /// Position of `source` in `sources`; unknown Filter Sets sort
    /// after every known source so a stale reference never jumps
    /// ahead of real ones.
    public func sourceOrderIndex(of source: FilterSource) -> Int {
        sources.firstIndex(of: source) ?? sources.count
    }

    public func contains(_ source: FilterSource) -> Bool {
        switch source {
        case .standard:
            return true
        case .filterSet(let id):
            return filterSet(withID: id) != nil
        }
    }
}

File: android/core/src/main/kotlin/com/sangwook/ptimer/core/exposure/FilterInventory.kt
// Copyright © 2026 Sangwook Han
// SPDX-License-Identifier: Apache-2.0

package com.sangwook.ptimer.core.exposure

import com.sangwook.ptimer.core.exposure.ExposureCalculator.Companion.STABILITY_EPSILON
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID
import kotlin.math.abs
import kotlin.math.log2
import kotlin.random.Random

/**
 * Stable identity of a user-owned Filter Set. Names and colors are
 * presentation metadata; only this id participates in stack references
 * and captured summaries. (iOS: `FilterSetID`.)
 */
data class FilterSetId(val rawValue: String) {
    companion object {
        fun generate(): FilterSetId = FilterSetId(UUID.randomUUID().toString())
    }
}

/**
 * Stable identity of one physical filter. Two items with equal names
 * and values remain two distinct physical filters. (iOS: `FilterItemID`.)
 */
data class FilterItemId(val rawValue: String) {
    companion object {
        fun generate(): FilterItemId = FilterItemId(UUID.randomUUID().toString())
    }
}

/**
 * Fixed, platform-neutral color palette for Filter Sets. Persisted by
 * [name] so both platforms map the same token to their own color
 * system. Duplicate colors across Filter Sets are allowed.
 * (iOS: `FilterSetColor`.)
 */
enum class FilterSetColor {
    red,
    orange,
    yellow,
    green,
    mint,
    teal,
    cyan,
    blue,
    indigo,
    purple,
    pink,
    brown;

    companion object {
        /** Fail-safe parse: an unknown/missing token restores as [blue]. */
        fun fromToken(token: String?): FilterSetColor =
            entries.firstOrNull { it.name == token } ?: blue

        /**
         * Random suggestion for a new Filter Set that differs from the
         * suggestion offered on the immediately preceding creation
         * opening. With one color excluded there are always eleven
         * candidates left, so the call never fails.
         */
        fun suggestion(excluding: FilterSetColor?, random: Random = Random.Default): FilterSetColor =
            entries.filter { it != excluding }.random(random)
    }
}

/**
 * Input unit of a registered filter value. Conversion to canonical
 * stops: Stops unchanged, `OD / 0.3`, `log2(ND factor)`. Persisted by
 * [name]; the tokens are part of the on-disk contract.
 * (iOS: `FilterValueUnit`.)
 */
@Serializable
enum class FilterValueUnit {
    @SerialName("stops")
    stops,

    @SerialName("opticalDensity")
    opticalDensity,

    @SerialName("filterFactor")
    filterFactor;

    companion object {
        fun fromToken(token: String?): FilterValueUnit? = entries.firstOrNull { it.name == token }
    }
}

/**
 * A registered Fixed or GND filter value as the user entered it — the
 * original value and unit are preserved for display, while calculation
 * consumes [canonicalStops]. (iOS: `FilterRegisteredValue`.)
 */
data class FilterRegisteredValue(val value: Double, val unit: FilterValueUnit) {

    /**
     * Canonical stops, or `null` when the value cannot be registered:
     * non-finite, zero or negative stops, or above the 30-stop cap
     * (FILTER-ITEM-004). An ND factor that exactly matches a commercial
     * Standard label ([NDCommercialFactorMapping]) takes that label's
     * ladder value — ND1000 is exactly 10 stops; any other factor uses
     * `log2` without snapping.
     */
    val canonicalStops: Double?
        get() {
            if (!value.isFinite()) return null
            val stops = when (unit) {
                FilterValueUnit.stops -> value
                FilterValueUnit.opticalDensity -> value / 0.3
                FilterValueUnit.filterFactor ->
                    NDCommercialFactorMapping.canonicalStops(matchingCommercialFactor = value)
                        ?: if (value > 0) log2(value) else Double.NEGATIVE_INFINITY
            }
            val withinCap = stops <= ExposureScale.MAXIMUM_WHOLE_ND_STOPS + STABILITY_EPSILON
            return if (stops.isFinite() && stops > 0 && withinCap) stops else null
        }

    val isValid: Boolean get() = canonicalStops != null
}

/**
 * Parses user-entered decimal text for filter registration. Accepts
 * either `.` or `,` as the decimal separator so locale keyboards,
 * pasted text, and hardware keyboards normalize to one rule
 * (FILTER-CPL-003). (iOS: `FilterDecimalInput`.)
 */
object FilterDecimalInput {
    private val decimalPattern = Regex("^[0-9]+(\\.[0-9]+)?$|^\\.[0-9]+$")
    private val cplFieldPattern = Regex("^[0-9](\\.[0-9])?$")

    internal fun normalizedText(text: String): String = text.trim().replace(',', '.')

    /**
     * General decimal parse (Fixed / GND values). `null` for empty or
     * non-numeric text.
     */
    fun parseDecimal(text: String): Double? {
        val normalized = normalizedText(text)
        if (!decimalPattern.matches(normalized)) return null
        return normalized.toDoubleOrNull()
    }

    /**
     * CPL exposure-loss field parse: one integer digit, at most one
     * fractional digit, in the closed range 0.1–9.9 (FILTER-CPL-002).
     */
    sealed class CplFieldParse {
        /** Blank text — the field omits a choice. */
        data object Empty : CplFieldParse()

        data class Value(val stops: Double) : CplFieldParse()

        data object Invalid : CplFieldParse()
    }

    fun parseCplField(text: String): CplFieldParse {
        val normalized = normalizedText(text)
        if (normalized.isEmpty()) return CplFieldParse.Empty
        if (!cplFieldPattern.matches(normalized)) return CplFieldParse.Invalid
        val value = normalized.toDoubleOrNull() ?: return CplFieldParse.Invalid
        return if (CplExposureLossChoices.isValidChoice(value)) {
            CplFieldParse.Value(value)
        } else {
            CplFieldParse.Invalid
        }
    }
}

/**
 * The three user-editable CPL exposure-loss choices. A field may be
 * empty (omits a choice); at least one field must hold a valid value
 * (FILTER-CPL-001/002). (iOS: `CPLExposureLossChoices`.)
 */
class CplExposureLossChoices(fields: List<Double?>) {

    /** Exactly three slots; `null` is an empty field. */
    val fields: List<Double?> = List(FIELD_COUNT) { fields.getOrNull(it) }

    val isValid: Boolean
        get() {
            val present = fields.filterNotNull()
            return present.isNotEmpty() && present.all { isValidChoice(it) }
        }

    /**
     * Distinct valid choices in ascending order — the rows the shooting
     * wheel offers for this item. Duplicates appear once.
     */
    val shootingChoices: List<Double>
        get() {
            val seen = ArrayList<Double>()
            for (value in fields.filterNotNull()) {
                if (!isValidChoice(value)) continue
                if (seen.none { abs(it - value) <= STABILITY_EPSILON }) seen.add(value)
            }
            return seen.sorted()
        }

    override fun equals(other: Any?): Boolean =
        this === other || (other is CplExposureLossChoices && other.fields == fields)

    override fun hashCode(): Int = fields.hashCode()

    override fun toString(): String = "CplExposureLossChoices(fields=$fields)"

    companion object {
        const val FIELD_COUNT: Int = 3
        private val range = 0.1..9.9

        /** Shipping defaults for a new CPL item: 1, 1.5, and 2 stops. */
        val defaults = CplExposureLossChoices(listOf(1.0, 1.5, 2.0))

        /**
         * A choice is valid when it is finite, inside 0.1–9.9, and has at
         * most one fractional digit (so `1.25` is rejected).
         */
        fun isValidChoice(value: Double): Boolean {
            if (!value.isFinite() || value !in range) return false
            val tenths = value * 10
            return abs(tenths - tenths.swiftRounded()) <= STABILITY_EPSILON
        }
    }
}

/**
 * Per-shot GND calculation mode (FILTER-GND-001). Persisted by [name].
 * (iOS: `GNDCalculationMode`.)
 */
enum class GndCalculationMode {
    /** The GND is mounted and recorded; it contributes 0 stops. */
    recordOnly,

    /** The GND contributes its complete registered value. */
    applyFullValue;

    companion object {
        fun fromToken(token: String?): GndCalculationMode? = entries.firstOrNull { it.name == token }
    }
}

/**
 * Behavior kind of a physical item — chosen explicitly by the user,
 * never inferred from the name (FILTER-ITEM-003). Persisted by [name].
 * (iOS: `FilterItemKind`.)
 */
@Serializable
enum class FilterItemKind {
    @SerialName("fixed")
    fixed,

    @SerialName("cpl")
    cpl,

    @SerialName("gnd")
    gnd;

    companion object {
        fun fromToken(token: String?): FilterItemKind? = entries.firstOrNull { it.name == token }
    }
}

/** (iOS: `FilterItemBehavior`.) */
sealed class FilterItemBehavior {
    data class Fixed(val value: FilterRegisteredValue) : FilterItemBehavior()
    data class Cpl(val choices: CplExposureLossChoices) : FilterItemBehavior()
    data class Gnd(val value: FilterRegisteredValue) : FilterItemBehavior()

    val kind: FilterItemKind
        get() = when (this) {
            is Fixed -> FilterItemKind.fixed
            is Cpl -> FilterItemKind.cpl
            is Gnd -> FilterItemKind.gnd
        }

    /**
     * Registered full-density value for Fixed and GND items; CPL items
     * carry choices instead.
     */
    val registeredValue: FilterRegisteredValue?
        get() = when (this) {
            is Fixed -> value
            is Gnd -> value
            is Cpl -> null
        }

    val isValid: Boolean
        get() = when (this) {
            is Fixed -> value.isValid
            is Gnd -> value.isValid
            is Cpl -> choices.isValid
        }
}

/**
 * One physical filter with a stable id. Equal names, kinds, units, and
 * values do not make two items the same item (FILTER-ITEM-002).
 * (iOS: `FilterItem`.)
 */
data class FilterItem(
    val name: String,
    val behavior: FilterItemBehavior,
    val id: FilterItemId = FilterItemId.generate(),
) {
    val isWellFormed: Boolean get() = name.trim().isNotEmpty() && behavior.isValid
}

/**
 * A user-owned, ordered group of physical filters with a stable id, a
 * non-empty name, and a required color (FILTER-SET-002).
 * (iOS: `FilterSet`.)
 */
data class FilterSet(
    val name: String,
    val color: FilterSetColor,
    val items: List<FilterItem> = emptyList(),
    val id: FilterSetId = FilterSetId.generate(),
) {
    val isWellFormed: Boolean get() = name.trim().isNotEmpty()

    fun item(withId: FilterItemId): FilterItem? = items.firstOrNull { it.id == withId }
}

/**
 * The complete user inventory: Filter Sets in user-defined display
 * order. Standard is a fixed built-in source that always precedes every
 * Filter Set (FILTER-SET-004). (iOS: `FilterInventory`.)
 */
data class FilterInventory(val filterSets: List<FilterSet> = emptyList()) {

    fun filterSet(withId: FilterSetId): FilterSet? = filterSets.firstOrNull { it.id == withId }

    fun item(withId: FilterItemId): Pair<FilterSet, FilterItem>? {
        for (filterSet in filterSets) {
            val item = filterSet.item(withId)
            if (item != null) return filterSet to item
        }
        return null
    }

    /**
     * Filter Sources in selection order: Standard, then Filter Sets in
     * their user-defined order.
     */
    val sources: List<FilterSource>
        get() = listOf(FilterSource.Standard) + filterSets.map { FilterSource.FilterSet(it.id) }

    /**
     * Position of [source] in [sources]; unknown Filter Sets sort after
     * every known source so a stale reference never jumps ahead of real
     * ones.
     */
    internal fun sourceOrderIndex(source: FilterSource): Int =
        sources.indexOf(source).takeIf { it >= 0 } ?: sources.size

    fun contains(source: FilterSource): Boolean = when (source) {
        is FilterSource.Standard -> true
        is FilterSource.FilterSet -> filterSet(source.id) != null
    }

    companion object {
        val empty = FilterInventory()
    }
}
